compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -1123,7 +1125,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 }));
                 let qpath = self.lower_qpath(
                     lhs.id,
-                    &None,
+                    qself,
```

Comment:
Other code here in `fn destructure_assign` needs to be kept symmetric with whatever is supported in the pattern syntax.

So if `<A>::B(..)` or `<A>::B { .. }` are supported in patterns, then equivalent expression should be accepted in destructuring assignments as well (preferably with a test).

Reply:
@petrochenkov it's easy to update the calls in `destructure_assign` to properly resolve `qself`, but I've run into an issue when checking that a lhs is [assignable](https://github.com/rust-lang/rust/blob/a93699f20a433797a4b84787b9652300dd7b2ad2/compiler/rustc_typeck/src/check/expr.rs#L732). The call into `is_syntactic_place_expr` returns false because `<A>::B(..)` is an `ExprKind::Call`, and I'm not sure of disambiguating between a regular function call and a tuple struct. Thoughts on this?